### PR TITLE
loosen `skipchars` signature

### LINKED ISSUE #21354: Loosen `skipchars` to IO?
`skipchars` currently only works for `IOStream`, but its implementation should work on any `IO` type that is backed by some underlying UTF-8 data and supports negative `skip`. Would it make sense to loosen the `IOStream` here to `IO`?

https://github.com/JuliaLang/julia/blob/90cfb8292c5d8cdd25eb4b4653c03f6d382b42b1/base/iostream.jl#L324

```julia
function skipchars(io::IOStream, pred; linecomment=nothing)
    while !eof(io)
        c = read(io, Char)
        if c === linecomment
            readline(io)
        elseif !pred(c)
            skip(io, -codelen(c))
            break
        end
    end
    return io
end
```

### LINKED ISSUE #22168: UniformScaling should use parentheses if the scalar is complex
entering `im*I+I` is shown as `1+1im*I` though it calculates as `(1+1im)*I` thus i guess that's some sort of wrong display algorithm.

(Since i've no clue of committing rules i thought that'd be a really easy to fix problem for you and can happen pretty much instant within your normal workflow ^^')

### PR DESCRIPTION
**Pull request by nsmith5**
fixes #21354

### CODE REVIEW

**Comment by KristofferC** [comment]
Think it fixes another issue than what you linked. Would be nice if you could add a test that tests this with e.g. an `IOBuffer`.

**Comment by nsmith5** [comment]
Good point. I'll add some tests. Probably should be moved into `io.jl` instead of `iostream.jl` as well given the change of signature. I'll close this and open another pull request with something more polished later...

**Comment by ararslan** [comment]
Typically it's preferable to update an existing pull request rather than closing one and opening another. Just an FYI for next time. Thanks for working on this!

**Comment by nsmith5** [comment]
Ok, thanks for the heads up!

**Comment by nsmith5** [comment]
Perhaps I'll just reopen the PR then and add "WIP" so people know I'm still working on this?

**Comment by nsmith5** [comment]
Hmm, seems Github was down for the mac test. Is there someway to trigger Travis again without a commit?

**Comment by ararslan** [comment]
I restarted the macOS build for the most recent commit.

File: base/io.jl [hunk 1]
```diff
@@ -645,3 +645,36 @@ ismarked(io::IO) = io.mark >= 0
...
+    skipchars(io::IO, predicate; linecomment=nothing)
+
+Skip forward in `io` until `predicate` returns `false`. If `linecomment`
+is defined, all characters after the `linecomment` character are ignored
+until the next line.
+
+```jldoctext
+julia> buf = IOBuffer("    text")
+IOBuffer(data=UInt8[...], readable=true, writable=false, seekable=true, append=false, size=8, maxsize=Inf, ptr=1, mar
+k=-1)
```
Comment (tkelman) [inline base/io.jl]: I don't think wrapping the output mid-keyword would pass if you rand the doctests, would it?

**Review by ararslan — APPROVED** [review]

File: base/io.jl [hunk 2]
```diff
@@ -645,3 +645,34 @@ ismarked(io::IO) = io.mark >= 0
 Commit all currently buffered writes to the given stream.
 """
 flush(io::IO) = nothing
+
+"""
+    skipchars(io::IO, predicate; linecomment=nothing)
```
Comment (tkelman) [inline base/io.jl]: there's a duplicate of this docstring now in `base/docs/helpdb/Base.jl` that should be removed

Comment (KristofferC) [inline base/io.jl]: Oh, I thought this was a brand new docstring. Should have realized..

Comment (nsmith5) [inline base/io.jl]: Oh my bad. I didn't see that either. How should I proceed here? Restore branch and remove one of the docstrings?

Comment (nsmith5) [inline base/io.jl]: Or just submit a new pull request?

Comment (KristofferC) [inline base/io.jl]: Just new PR

Comment (nsmith5) [inline base/io.jl]: Ok will do.

**Pull request merged**